Re-implement security module in python

rickheil opened the pull request:
This is stage 1 of improvements to the security model that I've been working on. 

This PR re-implements the security module script to use python instead of bash to make further improvements easier (other features need python!). It also changes the cache from a text file to a plist, which again will make life easier down the road.

Thanks and credit to @clburlison for suggesting and coding the plist piece of the python script!

rickheil commented:
Updated with stage 2: lists users who have access to ARD and SSH

Updated with stage 3: firmware password detection

rickheil commented:
Bugs fixed with open-to-everyone ARD and SSH - thank you steffan from slack for reporting!

Last to-do is the SIP and GateKeeper widgets.

gmarnin commented:
@rickheil I just pulled your wip branch then ran `sudo /bin/bash -c "$(curl -s http://127.0.0.1/munkireport-php/index.php?/install)"` (on 10.12.2) and got this traceback:

``` Munkireport: Running security.py
    Munkireport: security.py Error: Traceback (most recent call last):
  File "/usr/local/munki/preflight.d/security.py", line 196, in <module>
    main()
  File "/usr/local/munki/preflight.d/security.py", line 188, in main
    result.update({'ard_users': ard_access_check()})
  File "/usr/local/munki/preflight.d/security.py", line 118, in ard_access_check
    plist = plistlib.readPlist(plist_path)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 78, in readPlist
    rootObject = p.parse(pathOrFile)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 406, in parse
    parser.ParseFile(fileobj)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 1, column 9
WARNING: Munkireport: security.py return code: 1
```

rickheil commented:
@gmarnin thank you for testing. Steffan pointed this out on slack, error is because that plist is binary format and I'm using plistlib. I'm going to be changing it around to use FoundationPlist to avoid that problem.

Any other bugs you run into please let me know.

gmarnin commented:
Ok, I'll retest when that is fixed as I need data to test with.

rickheil commented on app/modules/security/scripts/install.sh:
```diff
@@ -1,8 +1,8 @@
-MODULESCRIPT="security.sh"
-MODULE_CACHE_FILE="security.txt"
+MODULESCRIPT="security.py"
+MODULE_CACHE_FILE="security.plist"
```
fixed in c3a2a2c87

rickheil replied:
Sorry, c3a2a2c8

rickheil commented on app/modules/security/scripts/uninstall.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
 
-rm -f "${MUNKIPATH}preflight.d/security.sh"
```
Fixed in 5c27ca60

rickheil commented on app/modules/security/security_model.php:
```diff
@@ -32,38 +32,19 @@ public function __construct($serial = '')
-            'SIP: ' => 'sip');
+	require_once(APP_PATH . 'lib/CFPropertyList/CFPropertyList.php');
+	$parser = new CFPropertyList();
+	$parser->parse($data);
```
Stole some logic from the Bluetooth module, with 39ce9739 it will parse both old (txt) and new (plist).

rickheil commented:
Ready for additional testing - addressed all of the comments in the review as well as a few bugs that were brought up on Slack.

rickheil commented:
Added a few bug fixes with the firmware password checks, and support for systems that only have SSH enabled from CLI with systemsetup and not the preference pane. 

Also added in a firmware password count widget, and a basic security report (right now just the security module widgets).

Testing completed on 10.7 through 10.12 and didn't find any bugs.

bryanzak commented:
Is munkireport dropping support for 10.6?

Sent from my iPhone

> On Jan 3, 2017, at 12:41 PM, Rick Heil <notifications@github.com> wrote:
> 
> Added a few bug fixes with the firmware password checks, and support for systems that only have SSH enabled from CLI with systemsetup and not the preference pane.
> 
> Also added in a firmware password count widget, and a basic security report (right now just the security module widgets).
> 
> Testing completed on 10.7 through 10.12 and didn't find any bugs.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rickheil commented:
@bryanzak if you have 10.6 around, I would very much welcome testing on it and any bugs you find. I don't have any 10.6 machines in my environment so I don't keep a 10.6 VM around.

bochoven commented on assets/locales/fr.json:
```diff
@@ -584,7 +584,15 @@
     "security": "Sécurité",
     "gatekeeper": "Etat Gatekeeper",
-    "sip": "Etat SIP"
+    "gatekeeper_active": "Gatekeeper Active",
```
Please don't update the french localisation file unless you want to translate to french. Munkireport fills in the missing phrases with the default language (english). Leaving the phrases out of the localisation file makes it easier for the translators to spot missing items.

rickheil replied:
Got it, I misunderstood a discussion on how to best do this on slack. I'll update to remove the added bits to both fr and de that I made.

rickheil replied:
Fixed with 68fe9ee and 7502d61
